cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+    
+    PromiseResult res = PROMISE_RESULT_CHANGE;
+    
+    const char *package_install_command = NULL;
+    if (type == PACKAGE_TYPE_FILE)
+    {
+        package_install_command = "file-install";
+    }
+    else if (type == PACKAGE_TYPE_REPO)
+    {
+        package_install_command = "repo-install";
+    }
+    else
+    {
+        /* If we end up here something bad has happened. */
+        ProgrammingError("Unsupported package type");
+    }
+    
+    Rlist *error_message = NULL;
+    if (ReadWriteDataToPackageScript(package_install_command, request, 
+                                     &error_message, wrapper) != 0)
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while communicating with "
+            "package module while installing package.");
+        res = PROMISE_RESULT_FAIL;
+    }
+    if (error_message)
+    {
+        ParseAndLogErrorMessage(error_message);
+        res = PROMISE_RESULT_FAIL;
+        RlistDestroy(error_message);
+    }
+    
+    free(request);
+    free(options_str);
+    
+    return res;
+}
+
+PromiseResult InstallPackage(Rlist *options, 
+        PackageType type, const char *package_to_install,
+        const char *version, const char *architecture,
+        const PackageModuleWrapper *wrapper)
+{
+    Log(LOG_LEVEL_DEBUG, "Installing package '%s'", package_to_install);
+             
+    char *ver = version ? 
+        StringFormat("Version=%s\n", version) : NULL;
+    char *arch = architecture ? 
+        StringFormat("Architecture=%s\n", architecture) : NULL;
+    char *request = NULL;
+    
+    PromiseResult res = PROMISE_RESULT_CHANGE;
+    
+    if (type == PACKAGE_TYPE_FILE)
+    {
+        request = StringFormat("File=%s\n%s%s",
+            package_to_install, ver ? ver : "", arch ? arch : "");
+    }
+    else if (type == PACKAGE_TYPE_REPO)
+    {
+        request = StringFormat("Name=%s\n%s%s",
+            package_to_install, ver ? ver : "", arch ? arch : "");
+    }
+    
+    res = InstallPackageGeneric(options, type, request, wrapper);
+    
+    free(request);
+    free(ver);
+    free(arch);
+    
+    return res;
+}
+
+PromiseResult FileInstallPackage(const char *package_file_path, 
+                                 const PackageInfo *info,
+                                 const NewPackages *policy_data,
+                                 const PackageModuleWrapper *wrapper,
+                                 int is_in_cache, enum cfopaction action)
+{
+    Log(LOG_LEVEL_DEBUG, "Installing file type package.");
+    
+    /* We have some packages matching file package promise in cache. */
+    if (is_in_cache == 1)
+    {
+        Log(LOG_LEVEL_VERBOSE, "Package exists in cache. Skipping installation.");
+        return PROMISE_RESULT_NOOP;
+    }
+    
+    PromiseResult res;
+     
+    if (action == cfa_warn || DONTDO)
+    {
+         Log(LOG_LEVEL_NOTICE, "Should install file type package: %s",
+             package_file_path);
+        res = PROMISE_RESULT_FAIL;
```

Comment:
I take it `DONTDO` is the dry-run flag.  I would have though `PROMISE_RESULT_NOOP` would be more apt there; but follow whatever other `DONTDO` cases do !